### PR TITLE
feat(dspy) PremAI python sdk

### PR DESCRIPTION
**Pull request by Anindyadeep**
This PR adds a integration of [premai](https://premai.io) [python sdk](https://github.com/premAI-io/prem-python-sdk) to dspy modules. 

Fixes issue: #1006

### CODE REVIEW

File: dsp/modules/premai.py
Comment (arnavsinghvi11) [inline dsp/modules/premai.py:13]: should this say Prem AI instead of Mistral AI?

File: dsp/modules/premai.py
Comment (arnavsinghvi11) [inline dsp/modules/premai.py:5]: add this import to the try/except block

File: dsp/modules/premai.py
Comment (arnavsinghvi11) [inline dsp/modules/premai.py:18]: we can actually remove this since we don't support module_specific logging anymore.

File: dsp/modules/premai.py
Comment (arnavsinghvi11) [inline dsp/modules/premai.py:92]: typo - change to `session_id`

File: dsp/modules/premai.py
Comment (arnavsinghvi11) [inline dsp/modules/premai.py:137]: would it be better to raise this error if the content field is not filled?

File: dsp/modules/premai.py
Comment (arnavsinghvi11) [inline dsp/modules/premai.py:84]: missing the line `self.history: list[dict[str, Any]] = []`

File: dsp/modules/premai.py
Comment (arnavsinghvi11) [inline dsp/modules/premai.py:83]: this is used to set a condition for how the prompts are referenced while inspecting the history of the LM. Can you check `lm.py` to see how this provider condition pertains to Prem AI? Feel free to follow this [part of the code](https://github.com/stanfordnlp/dspy/pull/944/files#diff-6bdc17664251e0662d9b19a69a5ebd5acd2252e4ecde2bdb6d89c460b1d022a6R64) to make the change?

**Review by arnavsinghvi11 — CHANGES_REQUESTED** [review]

**Comment by arnavsinghvi11** [comment]
Hi @Anindyadeep , thanks for the PR!

Added a few comments to resolve the failing build tests. 

Could you also add documentation for the PremAI LM to provide context as done for the LMs in [our documentation here](https://github.com/stanfordnlp/dspy/tree/main/docs/api/language_model_clients)?

Ready to merge after that!

File: dsp/modules/premai.py
Comment (Anindyadeep) [reply dsp/modules/premai.py:13]: Ahh yes, I used some of the code from Mistral AI for starters, thanks for pointing out. Will check it

File: dsp/modules/premai.py
Comment (Anindyadeep) [reply dsp/modules/premai.py:18]: So the problem, was I initially saw some implementation with other modules, and I also did in the same way. But, then all the pre-commit checks were failing. So I went to this route.

File: dsp/modules/premai.py
Comment (Anindyadeep) [reply dsp/modules/premai.py:13]: Done

File: dsp/modules/premai.py
Comment (Anindyadeep) [reply dsp/modules/premai.py:5]: This is not required, I removed `premai.errors`

File: dsp/modules/premai.py
Comment (Anindyadeep) [reply dsp/modules/premai.py:92]: Done

File: dsp/modules/premai.py
Comment (Anindyadeep) [reply dsp/modules/premai.py:137]: I mean, yes we can, but we took an assumption that, if `response.choices` exists, then this will also mean that, there should be some response that should exist. If not then there could be any internal error from Prem's side. So, for that we handled it in the above way where if this thing does not exist, then just return ""


what do you think?

File: dsp/modules/premai.py
Comment (Anindyadeep) [reply dsp/modules/premai.py:84]: Thanks for pointing this out, but I wonder, `self.history` is also been initialized in the abstract base class, so why we also define it here?

**Comment by Anindyadeep** [comment]
> Hi @Anindyadeep , thanks for the PR!
> 
> Added a few comments to resolve the failing build tests.
> 
> Could you also add documentation for the PremAI LM to provide context as done for the LMs in [our documentation here](https://github.com/stanfordnlp/dspy/tree/main/docs/api/language_model_clients)?
> 
> Ready to merge after that!

Hey thanks for doing the reviews, I will add the requested changes soon, also I would also like to discuss more on the pointers raised in replies.

**Comment by Anindyadeep** [comment]
Hey @arnavsinghvi11, I added some of the changes and feedbacks you provided, but for some reason, I am not able to pass the lint tests (with pre-commit), even for files like `lm.py` where problems are getting raised for lines I have not changed. So what to do on those cases?

File: dsp/modules/premai.py
Comment (Anindyadeep) [reply dsp/modules/premai.py]: Removed logger and replaced with print

**Comment by Anindyadeep** [comment]
Hi @arnavsinghvi11, it would great, if we can do one or two more iterations on this whenever you have some time. 

Thanks

File: dsp/modules/premai.py
Comment (arnavsinghvi11) [reply dsp/modules/premai.py:87]: defining the history here gives some flexibility with the typing (that is not done in the LM class). This behavior is maintained in most of the other LM classes ([gpt3.py](https://github.com/stanfordnlp/dspy/blob/0e595a70cd9d14b0db25252320aad11eae08fd72/dsp/modules/gpt3.py#L94), [Ollama](https://github.com/stanfordnlp/dspy/blob/main/dsp/modules/ollama.py), etc.).

Feel free to test the PremAI behavior with lm.inspect_history(n=1) on your end (where lm is defined as your dspy.PremAI model) as well.

File: dsp/modules/premai.py
Comment (arnavsinghvi11) [reply dsp/modules/premai.py]: My comment as more on if `response.choices[0].message.content` is empty, this seems like it would be some incomplete or unexpected response right? and that would be better to raise as a flag over returning `""`.

**Comment by arnavsinghvi11** [comment]
Hi @Anindyadeep , left a few comments. 


Please refer to these lines in the failing tests on how to resolve and push the changes

https://github.com/stanfordnlp/dspy/actions/runs/9067260753/job/25022498634?pr=1007#step:5:10

https://github.com/stanfordnlp/dspy/actions/runs/9067260753/job/25022499093?pr=1007#step:8:28

File: dsp/modules/premai.py
Comment (arnavsinghvi11) [inline dsp/modules/premai.py:11]: note that this error shouldn't be raised here. this causes the failing test. 

Instead, you should just define the error and should be triggered only when someone uses dspy.PremAI and doesn't have it installed.

Refer to [other LM providers](https://github.com/stanfordnlp/dspy/blob/0e595a70cd9d14b0db25252320aad11eae08fd72/dsp/modules/cohere.py#L10) on how to resolve this,

File: dsp/modules/premai.py
Comment (Anindyadeep) [reply dsp/modules/premai.py]: I see, that makes sense. Doing those changes

File: dsp/modules/premai.py
Comment (Anindyadeep) [reply dsp/modules/premai.py:11]: I see, yeah missed resolving this one, thanks for pointing out.

File: dsp/modules/premai.py
Comment (Anindyadeep) [reply dsp/modules/premai.py:85]: Resolved

File: dsp/modules/premai.py
Comment (Anindyadeep) [reply dsp/modules/premai.py]: Resolved, now all tests are running.

**Comment by Anindyadeep** [comment]
So, tests running now, and fixed some bugs, hopefully it should pass the ci workflow checks. For ruff, I got this:

```
vscode ➜ /workspaces/dspy (modules/premai) $ ruff check . --fix-only                                   
warning: The top-level linter settings are deprecated in favour of their counterparts in the `lint` section. Please update the following options in `pyproject.toml`:
  - 'extend-unsafe-fixes' -> 'lint.extend-unsafe-fixes'
warning: Docstring at dspy/teleprompt/mipro_optimizer.py:111:9 contains implicit string concatenation; ignoring...
warning: Docstring at dspy/teleprompt/mipro_optimizer.py:122:9 contains implicit string concatenation; ignoring...
```

Which are nothing related to files on this integration.

**Comment by arnavsinghvi11** [comment]
> Hi @arnavsinghvi11 , So, all tests passing now, and fixed some bugs, hopefully it should pass the ci workflow checks. For ruff, I got this:
> 
> ```
> vscode ➜ /workspaces/dspy (modules/premai) $ ruff check . --fix-only                                   
> warning: The top-level linter settings are deprecated in favour of their counterparts in the `lint` section. Please update the following options in `pyproject.toml`:
>   - 'extend-unsafe-fixes' -> 'lint.extend-unsafe-fixes'
> warning: Docstring at dspy/teleprompt/mipro_optimizer.py:111:9 contains implicit string concatenation; ignoring...
> warning: Docstring at dspy/teleprompt/mipro_optimizer.py:122:9 contains implicit string concatenation; ignoring...
> ```
> 
> Which are nothing related to files on this integration.

Hi @Anindyadeep , seems like those are just warnings. the ruff check passes within the PR so ready to merge whenever you are ready!

**Comment by Anindyadeep** [comment]
> > Hi @arnavsinghvi11 , So, all tests passing now, and fixed some bugs, hopefully it should pass the ci workflow checks. For ruff, I got this:
> > ```
> > vscode ➜ /workspaces/dspy (modules/premai) $ ruff check . --fix-only                                   
> > warning: The top-level linter settings are deprecated in favour of their counterparts in the `lint` section. Please update the following options in `pyproject.toml`:
> >   - 'extend-unsafe-fixes' -> 'lint.extend-unsafe-fixes'
> > warning: Docstring at dspy/teleprompt/mipro_optimizer.py:111:9 contains implicit string concatenation; ignoring...
> > warning: Docstring at dspy/teleprompt/mipro_optimizer.py:122:9 contains implicit string concatenation; ignoring...
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Which are nothing related to files on this integration.
> 
> Hi @Anindyadeep , seems like those are just warnings. the ruff check passes within the PR so ready to merge whenever you are ready!

Thats, awesome, thanks I will ping you once, we are ready for merge.

**Comment by Anindyadeep** [comment]
Hi @arnavsinghvi11, we are now ready get this merge.

**Comment by Anindyadeep** [comment]
Hi @arnavsinghvi11, sorry have to bother you again, but it would be awesome, if you could approve this one, so that it checks the CI and we ready to merge

thanks :)

**Comment by arnavsinghvi11** [comment]
Thanks @Anindyadeep !

**Comment by Anindyadeep** [comment]
Thanks @arnavsinghvi11 for the review :)

**Pull request merged**
